desktop-widgets/importgps.cpp

```diff
@@ -0,0 +1,302 @@
+
+#include <QFileDialog>
+#include <QProcess>
+#include <QDateTime>
+#include <QDate>
+
+/* Import dive coordinates from a GPS device and synchronise them with the dive profile information
+   of a dive computer. This file contains the infrastructure to:
+   1) Read a .GPX file from a GPS system.
+   2) Find the first gpx trackpoint that follows after the start of the dive.
+   3) Allow modification of the coordinates dependent on international time zone and
+              on differences in local time settings between the GPS and the dive computer.
+   4) Saving the coordinates into the Coordinates text box in the Dive Site Edit panel
+              and which which causes the map to show the location of the dive site. 
+   The structure coords is used to store critical information.  */
+ImportGPS::ImportGPS(QWidget *parent, QString fileName, class Ui::LocationInformation *LocationUI) : QDialog(parent),
+	fileName(fileName), LocationUI(LocationUI)
+{
+	ui.setupUi(this);
+	connect(ui.timeDiffEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeDiffEditChanged()));
+	connect(ui.timeZoneEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeZoneEditChanged()));
+	connect(ui.timezone_backwards, SIGNAL(toggled(bool)), this, SLOT(changeZoneBackwards()));
+	connect(ui.timezone_forward, SIGNAL(toggled(bool)), this, SLOT(changeZoneForward()));
+	connect(ui.diff_backwards, SIGNAL(toggled(bool)), this, SLOT(changeDiffBackwards()));
+	connect(ui.diff_forward, SIGNAL(toggled(bool)), this, SLOT(changeDiffForward()));
+	connect(ui.GPSbuttonBox, SIGNAL(clicked(QAbstractButton *)), this, SLOT(buttonClicked(QAbstractButton *)));
+	coords.settingsDiff_offset = 0;
+	coords.timeZone_offset = 0;
+	coords.lat = 0;
+	coords.lon = 0;
+	pixmapSize = (int) (ui.diveDateLabel->height() / 2);
+}
+
+void ImportGPS::buttonClicked(QAbstractButton *button)
+{
+	if (ui.GPSbuttonBox->buttonRole(button) == QDialogButtonBox::AcceptRole) {
+		// Write the coordinates in decimal degree format to the Coordinates QLineEdit of the LocationaInformationWidget UI:
+		LocationUI->diveSiteCoordinates->setText(QString::number(coords.lat) + ", " + QString::number(coords.lon));
+		LocationUI->diveSiteCoordinates->editingFinished();
+	} else {
+		close();
+	}
+}
+
+//  Read text from the present position in the file until
+//  the first 'delim' character is encountered.
+int ImportGPS::getSubstring(QFile *file, QString *bufptr, char delim)
+{
+	char c;
+	bufptr->clear();
+	do {
+		if (file->read(&c, 1) <= 0) // EOF
+			return 1;
+		if ( c == delim) break;
+		bufptr->append(QChar(c));
+	} while (c != delim);
+	return 0;
+}
+
+// Find the next occurence of a specified target GPX element in the file,
+// characerised by a "<xxx " or "<xxx>" character sequence.
+// 'target' specifies the name of the element searched for.
+// termc is the ending character of the element name search = '>' or ' '.
+int ImportGPS::findXmlElement(QFile *fileptr, QString target, QString *bufptr, char termc)
```

Comment:
Hmm.
Why are we manually parsing XML? 
Our CMake file looks like we link against libxml2.0 which suggests we should be able to parse the track with that?

Reply:
In fact, the dive log files in xml are parsed using a much more powerful framework than this gpx interpretation. However, when extracting a small amount of very specific information from xml it is, in my opinion, overkill to use one of the larger xml parsing frameworks. The existing code is pretty compact.

Reply:
Agree overkill if we were including it just for this. But if the library is already there why not just use it?

Reply:
I'll let @willemferguson speak for himself - for me it's often that simply parsing XML as strings seems "approachable" and dealing with all the machinations of these frameworks is often painful.

@buxtonpaul are you offering to convert Willem's code to using either libxml2 or the Qt XML functions? I don't want to make this a requirement for Willem before his code is merged...

@willemferguson one thing I really would like to have (and this __should__ be a separate commit) is a dive log XML and corresponding `.gpx` file so that I am able to actually test this :-)

And of course, this code should have been structured in a way that it is clearly separated into core and UI so that it can be called from mobile as well (another thing that I should have looked for earlier... I'm not doing a good job as maintainer these days, I'm afraid). So I guess I am offering to take care of that... :-(

Reply:
Keeping Core and UI separate:
As is, the core and UI are reasonable separate. I created the coords structure with the aim to keep the core separate; otherwise the members of coords could have been defined within the object definition in the header file. BUT, some of the values in the coords structure are obtained from the UI.